From: Julian Wiedmann <jwi@linux.vnet.ibm.com>
Subject: Subject: [PATCH] af_iucv: enable control sends in case of SEND_SHUTDOWN
Patch-mainline: v4.10-rc1
Git-commit: 4e0ad32216910905d00a1228379163c6e5057dc4
References: bnc#1085513, LTC#165135

Description:  af_iucv: send control flags on shutdown socket
Symptom:      af_iucv connections over HiperSockets stall.
Problem:      Once an af_iucv socket that uses HIPER Transport is shutdown for
              sending, it also no longer sends out receive window updates.
              These are required by protocol, so at some point the other end
              of the connection will stall because it can't send further data.
Solution:     For sending control flags, temporarily switch off the
              SEND_SHUTDOWN state on a socket.
Reproduction: af_iucv stress test

Upstream-Description:

              Subject: [PATCH] af_iucv: enable control sends in case of SEND_SHUTDOWN

              If a socket program has shut down the socket for sending, it can still
              receive an undetermined number of packets. The AF_IUCV protocol for
              HIPER transport requires sending of a WIN flag from time to time
              from the receiver to the sender, otherwise the peer cannot continue
              sending. That means sending of control flags must still work, even
              though the AF_IUCV socket is shutdown for sending data.
              sock_alloc_send_skb() returns with error EPIPE, if socket sk_shutdown
              is SEND_SHUTDOWN. Thus this patch temporarily removes the send
              shutdown attribute from the socket to enable transfer of control
              flags.

              Signed-off-by: Ursula Braun <ubraun@linux.vnet.ibm.com>
              Signed-off-by: David S. Miller <davem@davemloft.net>


Signed-off-by: Julian Wiedmann <jwi@linux.vnet.ibm.com>
Acked-by: Johannes Thumshirn <jthumshirn@suse.de>
---
 net/iucv/af_iucv.c |   10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

--- a/net/iucv/af_iucv.c
+++ b/net/iucv/af_iucv.c
@@ -475,19 +475,27 @@ static void iucv_sever_path(struct sock
 	}
 }
 
-/* Send FIN through an IUCV socket for HIPER transport */
+/* Send controlling flags through an IUCV socket for HIPER transport */
 static int iucv_send_ctrl(struct sock *sk, u8 flags)
 {
 	int err = 0;
 	int blen;
 	struct sk_buff *skb;
+	u8 shutdown = 0;
 
 	blen = sizeof(struct af_iucv_trans_hdr) + ETH_HLEN;
+	if (sk->sk_shutdown & SEND_SHUTDOWN) {
+		/* controlling flags should be sent anyway */
+		shutdown = sk->sk_shutdown;
+		sk->sk_shutdown &= RCV_SHUTDOWN;
+	}
 	skb = sock_alloc_send_skb(sk, blen, 1, &err);
 	if (skb) {
 		skb_reserve(skb, blen);
 		err = afiucv_hs_send(NULL, sk, skb, flags);
 	}
+	if (shutdown)
+		sk->sk_shutdown = shutdown;
 	return err;
 }
 
